src/test/librbd/io/test_mock_ImageRequest.cc

```diff
@@ -35,11 +35,24 @@ struct MockTestJournal : public MockJournal {
   MOCK_METHOD2(commit_io_event, void(uint64_t, int));
 };
 
+struct MockTestPolicy {
+  bool append_disabled() const {
+    return false;
```

Comment:
All other tests with imagectx.journal != nullptr would stall if I set journal_policy as a variable. I think it was enough to just have the variable and then override is_journal_appending. Maybe I'm missing something here but this way I was able to make it clear all tests.

Reply:
What if you don't bother with `journal_policy` at all?
```
diff --git a/src/test/librbd/io/test_mock_ImageRequest.cc b/src/test/librbd/io/test_mock_ImageRequest.cc
index b37c65cdf9e..fd500043721 100644
--- a/src/test/librbd/io/test_mock_ImageRequest.cc
+++ b/src/test/librbd/io/test_mock_ImageRequest.cc
@@ -35,24 +35,11 @@ struct MockTestJournal : public MockJournal {
   MOCK_METHOD2(commit_io_event, void(uint64_t, int));
 };

-struct MockTestPolicy {
-  bool append_disabled() const {
-    return false;
-  }
-
-  bool journal_disabled() const {
-    return false;
-  }
-
-  void allocate_tag_on_lock(Context *on_finish);
-};
-
 struct MockTestImageCtx : public MockImageCtx {
   MockTestImageCtx(ImageCtx &image_ctx) : MockImageCtx(image_ctx) {
   }

   MockTestJournal* journal;
-  MockTestPolicy* journal_policy;
 };

 } // anonymous namespace
@@ -404,8 +391,6 @@ TEST_F(TestMockIoImageRequest, EmptyDiscardJournalAppendEnabled) {
   ictx->discard_granularity_bytes = 32;

   MockTestImageCtx mock_image_ctx(*ictx);
-  librbd::MockTestPolicy policy;
-  mock_image_ctx.journal_policy = &policy;
   MockTestJournal mock_journal;
   mock_image_ctx.journal = &mock_journal;

@@ -438,8 +421,6 @@ TEST_F(TestMockIoImageRequest, PartialDiscardJournalAppendEnabled) {
   ictx->discard_granularity_bytes = 0;

   MockTestImageCtx mock_image_ctx(*ictx);
-  librbd::MockTestPolicy policy;
-  mock_image_ctx.journal_policy = &policy;
   MockTestJournal mock_journal;
   mock_image_ctx.journal = &mock_journal;

@@ -473,8 +454,6 @@ TEST_F(TestMockIoImageRequest, TailDiscardJournalAppendEnabled) {
   ictx->discard_granularity_bytes = 2 * ictx->layout.object_size;

   MockTestImageCtx mock_image_ctx(*ictx);
-  librbd::MockTestPolicy policy;
-  mock_image_ctx.journal_policy = &policy;
   MockTestJournal mock_journal;
   mock_image_ctx.journal = &mock_journal;

@@ -508,8 +487,6 @@ TEST_F(TestMockIoImageRequest, DiscardGranularityJournalAppendEnabled) {
   ictx->discard_granularity_bytes = 32;

   MockTestImageCtx mock_image_ctx(*ictx);
-  librbd::MockTestPolicy policy;
-  mock_image_ctx.journal_policy = &policy;
   MockTestJournal mock_journal;
   mock_image_ctx.journal = &mock_journal;

```

Reply:
My bad,

I was setting

bool appending = false;
bool is_journal_appending() {
  return appending;
}

in TestImageCtx which made tests stall. I would assume that it would work?

Reply:
Try applying just the diff that I pasted -- I think it should work.

Reply:
I'm quite certain that I tested this out before after mocking the policy. But yes, your patch worked fine :)